src/libawkward/Content.cpp

```diff
@@ -1348,6 +1348,16 @@ namespace awkward {
     return out;
   }
 
+  ContentPtr
+  Content::to_gpu(kernel::Lib ptr_lib) const{
+    return nullptr;
+  }
+
+  ContentPtr
+  Content::to_cpu() const{
+    return nullptr;
+  }
+
```

Comment:
Is this a default implementation: if a `Content` subclass doesn't override these, they get null-pointers? That's very dangerous! It would be much better if this function (`to` or `copy_to`, see above) is pure virtual:

```c++
const ContentPtr
  copy_to(kernel::Lib ptr_lib) const = 0;
```

in Content.h. Then the compiler forces us to implement it everywhere, rather than getting segfaults at runtime (which are hard to debug).

Reply:
I originally intended to do that. But I am stuck at understanding `VirtualArray` which inherits `Content`. Would a `copy_to` function for `VirtualArray` just be a pass-through?

Reply:
That is a good question, one I hadn't considered. At least at first, a VirtualArray has no control over where the array will be created because it's created by a Python function (when using Awkward from Python; otherwise one has to override a callback, as in your .cpp tests).

I suppose we could add a `kernel::Lib` enum to the VirtualArray class itself and then as soon as a new array is created, it is copied. This immediate "copy to" would have to happen in `VirtualArray::array`, wrapping the output of `generate_and_check()`.

https://github.com/scikit-hep/awkward-1.0/blob/62d79bc3b8b2f38c472856e7b884f36b1eb95076/src/libawkward/array/VirtualArray.cpp#L272-L285

But it's the caching that makes this tricky. VirtualArrays call a function to produce an array but do not store the results—they would keep calling their function, re-producing their array every time if they don't have a cache to put it in. This cache is a MutableMapping on the Python side and has that kind of interface in C++, too. (The old version of this would attach the generated array permanently to the VirtualArray object, but that was a problem because users couldn't easily make them go out of scope. By putting this all in a cache that the user controls, users can manage memory by clearing that cache object, or using one with LRU eviction rules.)

The problem is that a cached copy of the array might be on the "wrong" device (main memory or GPU) for a VirtualArray of a given `cache_key` (string). We don't want to wrap the `cache_.get()->get(cache_key())` in the above function with a "copy to" because that would repeatedly copy it from main memory to the GPU (or vice versa) every time it is accessed. I think the right thing to do for VirtualArray is this:

   * If its `kernel::Lib` enum is cpu-kernels, then the only change in behavior is wrapping `generate_and_check()` with a "copy to main memory," which would be a pass-through if the generated array is already in main memory (which is usually the case).
   * If its `kernel::Lib` enum is something else, then we wrap `generate_and_check()` with a "copy to desired location," which might or might not be a pass-through, depending on whether the user was conscientious enough to write their generating function to agree with the location of the VirtualArray in which it is used (something that should not be assumed). The copy that we put in the cache is the copy in the desired location, and we use a "fully qualified cache key," the regular `cache_key()` suffixed by a colon and the name of the desired location, such as `"my-super-awesome-cache-key:cuda"`.
   * When retrieving an array from cache (and `kernel::Lib` is not cpu-kernels), we first check the fully qualified cache key, then check the regular cache key, and if both fail, generate again.

In summary, (1) cpu-kernels is special; it's the only library for which the fully qualified cache key is equal to the regular cache key. (2) The only copy of the array that gets cached is the one in the desired location, using a fully qualified cache key. (3) There are now two fallbacks when attempting to retrieve a key from cache, rather than one: first fully qualified cache key, then regular cache key, then regenerate. All of this is assuming that the cache exists, of course. (It's very simple when the cache does not exist!)

A smart user would keep a different cache for GPU arrays than main memory arrays because they're not interchangeable resources: freeing up main memory doesn't give you more GPU global memory, and vice-versa. We can't enforce that from here, but it would be one of the "best practices" that I'd talk about in tutorials. However, we can make the algorithm friendly to good practices: we don't want to put both the pre-copied and post-copied array into the same cache, since this could mix main memory and GPU in the same cache, or at least put two references to the same array in the cache (if it's already in the right place), which undermines the usefulness of an LRU algorithm. We do, however, want to fetch an array assuming that the cache might be mixed, because maybe the user isn't following best practices and re-copying an array from main memory to GPU is likely better than re-running the generator function (which usually comes from disk; much slower than the memory → GPU copy).

This slide should be updated for GPUs!

![](https://slideplayer.com/slide/3360397/12/images/14/Jim+Gray%E2%80%99s+Storage+Latency+Analogy%3A+How+Far+Away+is+the+Data.jpg)

Reply:
```c
const ContentPtr
  VirtualArray::array() const {
    ContentPtr out(nullptr);
    if (cache_.get() != nullptr) {
      if(ptr_lib_ == kernel::Lib::cuda_kernels) {
        if(cache_key().find(":cuda") != std::string::npos) {
          out = cache_.get()->get(cache_key());
        }
        else {
          out = copy_to(kernel::Lib::cuda_kernels);
        }
      }
      out = cache_.get()->get(cache_key());
    }
    if (out.get() == nullptr) {
      out = copy_to(ptr_lib_);
    }
    if (cache_.get() != nullptr) {
      if(ptr_lib_ == kernel::Lib::cpu_kernels) {
        cache_.get()->set(cache_key(), out);
      }
      else if(ptr_lib_ == kernel::Lib::cuda_kernels){
        cache_.get()->set(cache_key() + ":cuda", out);
      }
    }
    return out;
  }
```
and `copy_to` is just:

```c
ContentPtr
  VirtualArray::copy_to(kernel::Lib ptr_lib) const {
      return generator_.get()->generate_and_check()->copy_to(ptr_lib);
  }
```
Does this make sense? Especially this part:
> When retrieving an array from cache (and kernel::Lib is not cpu-kernels), we first check the fully qualified cache key, then check the regular cache key, and if both fail, generate again.

Reply:
No, you shouldn't be expecting the `cache_key` to have `":cuda"` in it (and here, you're not controlling for where the substring is located; it should be at the end of the string), you should create a new key that is the user-provided `cache_key` plus the suffix `":cuda"` at the end. In other words, you _make_ the fully-qualified key; you don't expect to _get_ one.

Also, you can do this without any inline special cases on `ptr_lib_`. Suppose you have a function

```c++
const std::string
kernel::fully_qualified_cache_key(const std::string& cache_key, kernel::Lib ptr_lib) {
    switch (ptr_lib) {
    case kernel::Lib::cuda_kernels:
        return cache_key + std::string(":cuda");
    default:
        return cache_key;
    }
}
```

then you can first ask if the cache has a fully qualified key (which can be returned as-is), then ask if it has a plain `cache_key` (which needs to go through its `copy_to` and be added to the cache with a fully qualified key), then generate a new array as the second fallback. Newly generated arrays should go through their `copy_to` and only be stored in the cache with fully qualified keys.

The VirtualArray's `copy_to` (which is different from the generated array's own `copy_to`, just to be clear!) should not be calling the generator at all, even through `array()`. What `copy_to` means for a VirtualArray is to make it a VirtualArray-on-GPU (`copy_to` doesn't change array node type), which might still be ungenerated. It should copy the VirtualArray node but give it a new `ptr_lib` tag so that when arrays are asked for, they will end up on the appropriate device.

In summary,

   * VirtualArray has a new `ptr_lib_` tag to indicate what device its arrays will eventually end up on.
   * `VirtualArray::copy_to` just sets the value of that tag in a new copy of the VirtualArray node (lightweight to copy; it's just a bunch of references).
   * Every other `Content` subclass's `copy_to` actually copies (recursively) if the desired `kernel::Lib` is different from the actual one (for each node in the tree, recursively).
   * The `VirtualArray::array()` function needs to be `ptr_lib_` aware, so that it calls `copy_to` on any newly generated arrays to ensure that they're on the VirtualArray's own `ptr_lib_`, saves the final array in the cache with a fully qualified key, and checks the cache with two fallbacks instead of just one: fully qualified key (return what you've got because it's already in the right place), unqualified key (call its `copy_to` to ensure it's in the right place before saving it back to the cache with a fully qualified key), and generate anew.

Reply:
This would be good to test out in the `studies` directory because it's the logic of caching that you're concerned about, and it can be replicated in Python without needing much preexisting code. You can just use a dict for a cache and write up a simple VirtualArray class, then try it out in a variety of situations to see if it's making sense.

Unfortunately, I don't have a good starting implementation to use as a guide: https://awkward-array.readthedocs.io/en/latest/ak.layout.VirtualArray.html But maybe you can fill in the missing methods with the implementation before `ptr_lib_`, see how it works, and then add `ptr_lib_` to see if your algorithm is still sensible.

"Sensible" means that you should avoid filling the cache with two copies of the same array, only the one on the desired device, but you can't expect users to generate them in the right place, and they might use the same cache for CPU and GPU arrays, so it should deal with that in a way that copies as little as possible. You don't need anything realistic for the non-VirtualArray types; just a dummy class tagged with a `ptr_lib_`.

Reply:
I think I understood what `VirtualArrays` do but I do have some questions as to how they work internally(for eg. How is the cache_key generated, in what format is the cache stored etc) but I can ask you that some other time. I did make the necessary changes to the `::array()` and I think that should be, what you were trying to explain to me in the comments above.

Reply:
The cache_key can be user-supplied, but often it's a process-unique number from

https://github.com/scikit-hep/awkward-1.0/blob/6e6de66326883c80c4a20058409becba7be7daaf/src/libawkward/virtual/ArrayCache.cpp#L10-L15

The cache itself is an abstract MutableMapping, but usually a dict or ArrayCacheLRU from Python. From C++, it's just anything that can store Content objects, keyed by a string; the implementation is not restricted.

Reply:
Yeah, that makes a lot of sense. Thanks!

Reply:
I don't think the current implementation of `VirtualArray::array` is correct, but it won't break on non-GPU cases. You should create an issue to revisit it, but we can put that off for a long time.